### PR TITLE
feat:support backup/sync ens name

### PR DESCRIPTION
**Pull request by qfrank**
RT

### CODE REVIEW

**Review by saledjenic — APPROVED** [review]

**Comment by igor-sirotin** [comment]
FYI: I've previously introduced a new `ens_usernames` table, which replaces `settings.usernames`: https://github.com/status-im/status-go/pull/3066

The old `settings.usernames` still works and maintained by status-go, but desktop uses the new table now.

cc @cammellos

**Comment by qfrank** [comment]
> FYI: I've previously introduced a new `ens_usernames` table, which replaces `settings.usernames`: #3066
> 
> The old `settings.usernames` still works and maintained by status-go, but desktop uses the new table now.
> 
> cc @cammellos

interesting, i have something to say, do we really need a table to store ens names? it won't have much records in this table i think, and `settings.usernames` is a BLOB type represented with JSON, we can extend it dynamicly, and table `settings` is very convinient to implement sync/backup now @cammellos @igor-sirotin @saledjenic @Samyoul @richard-ramos

**Comment by cammellos** [comment]
> > FYI: I've previously introduced a new `ens_usernames` table, which replaces `settings.usernames`: #3066
> > The old `settings.usernames` still works and maintained by status-go, but desktop uses the new table now.
> > cc @cammellos
> 
> interesting, i have something to say, do we really need a table to store ens names? it won't have much records in this table i think, and `settings.usernames` is a BLOB type represented with JSON, we can extend it dynamicly, and table `settings` is very convinient to implement sync/backup now @cammellos @igor-sirotin @saledjenic @Samyoul @richard-ramos

We tend to prefer tables over JSON blobs, migrating json blobs requires user level migrations and it's very opaque and difficult to extend, so unless there's strong reason to use JSON and a single field, we prefer separate tables

**Comment by qfrank** [comment]
@cammellos emm.. so now we need support sync/backup table `ens_usernames` as well?

**Comment by saledjenic** [comment]
> interesting, i have something to say, do we really need a table to store ens names?

There are pros and cons for both approaches. I personally prefer having a table and a single column for each data, that gives a granularity, a possibility to write highly customized queries with different views of the database and prepare data as you need on the db level with high speed and without additional coding.

> so now we need support sync/backup table ens_usernames as well?

We definitely need it, but since mobile app is still using `settings.usernames` I thought you wanted to sync only that and leave to somebody from the desktop team to handle `ens_usernames`.

**Comment by qfrank** [comment]
> We definitely need it, but since mobile app is still using settings.usernames I thought you wanted to sync only that and leave to somebody from the desktop team to handle ens_usernames.

i'm going to do it in this PR. @saledjenic  cc @igor-sirotin

**Comment by igor-sirotin** [comment]
Thanks @qfrank ❤️. I wasn't able to catch up previous week.
You can leave it for me as well, but I think you will implement it much faster, as you've been around it for awhile already

File: services/ens/database.go [hunk 1]
```diff
@@ -8,21 +8,29 @@ type Database struct {
 	db *sql.DB
 }
 
-type UsernameDetails struct {
+type UsernameDetail struct {
```
Comment (igor-sirotin) [inline services/ens/database.go:11]: Surely this is a minor change, but does it make sense? It's not just 1 detail inside 🤔

File: protocol/protobuf/sync_settings.proto [hunk 2]
```diff
@@ -31,6 +31,7 @@ message SyncSetting {
     DISPLAY_NAME = 13;
     BIO = 14;
     MNEMONIC_REMOVED = 15;
+    USERNAMES = 16;
```
Comment (igor-sirotin) [inline protocol/protobuf/sync_settings.proto]: Maybe `ENS_USERNAMES` would be better here?

**Review by igor-sirotin — APPROVED** [review]
LGTM, thank you! 👍

File: protocol/messenger.go [hunk 3]
```diff
@@ -2902,6 +2907,78 @@ func (m *Messenger) SyncBookmark(ctx context.Context, bookmark *browsers.Bookmar
...
+
+	rawMessage := common.RawMessage{
+		LocalChatID:         contactIDFromPublicKey(&m.identity.PublicKey),
+		Payload:             encodedMessage,
+		MessageType:         protobuf.ApplicationMetadataMessage_SYNC_ENS_USERNAME_DETAIL,
+		ResendAutomatically: true,
+	}
+
+	_, err = rawMessageHandler(ctx, rawMessage)
+	if err != nil {
```
Comment (saledjenic) [inline protocol/messenger.go]: @qfrank minor thing, you could just `return err` here.

File: services/ens/database.go [hunk 4]
```diff
@@ -8,21 +8,29 @@ type Database struct {
...
 	Username string `json:"username"`
 	ChainID  uint64 `json:"chainId"`
+	Clock    int64  `json:"clock"`
```
Comment (saledjenic) [inline services/ens/database.go]: `uint64` maybe?

**Review by saledjenic — APPROVED** [review]

File: services/ens/database.go (hunk 1 above)
Comment (qfrank) [reply services/ens/database.go:11]: make sense when something like `var usernameDetails []*UsernameDetail` comes to me

> It's not just 1 detail inside

if you mean username|chainId are details, IMO, username/chainId could belongs to UsernameDetail, each of them is part of UsernameDetail. @igor-sirotin

**Pull request merged**